// @packages/app/dist/assets/CloudConnectButton.1d7b3801.js
import { d as defineComponent, b_ as useSlots, o as openBlock, X as createElementBlock, y as createBaseVNode, s as renderSlot, c as createBlock, a2 as resolveDynamicComponent, V as normalizeClass, t as toDisplayString, e as unref, i as createCommentVNode, u as useI18n, g as gql, m as ref, a as computed, z as createVNode, w as withCtx, f as createTextVNode, C as ConnectIcon, h as _sfc_main$2, A as _sfc_main$3, F as _sfc_main$4, a3 as Fragment } from "./index.0fb50328.js";
import { U as UserIcon } from "./user-outline_x16.69344947.js";
const _hoisted_1 = { class: "rounded flex text-left w-full py-14px items-center" };
const _hoisted_2 = { class: "flex h-40px px-24px items-center" };
const _hoisted_3 = { class: "flex-grow h-auto border-gray-100 border-l-1px px-16px" };
const _hoisted_4 = { class: "font-normal text-sm text-gray-700 select-none" };
const _hoisted_5 = {
  key: 0,
  class: "flex px-16px items-center"
};
const _sfc_main$1 = /* @__PURE__ */ defineComponent({
  props: {
    description: { default: void 0 },
    icon: { default: void 0 },
    gray: { type: Boolean, default: false },
    bigHeader: { type: Boolean, default: false }
  },
  setup(__props) {
    const slots = useSlots();
    return (_ctx, _cache) => {
      return openBlock(), createElementBlock("div", _hoisted_1, [
        createBaseVNode("div", _hoisted_2, [
          renderSlot(_ctx.$slots, "icon", {}, () => [
            (openBlock(), createBlock(resolveDynamicComponent(__props.icon), { class: "h-24px w-24px" }))
          ])
        ]),
        createBaseVNode("div", _hoisted_3, [
          createBaseVNode("h2", {
            class: normalizeClass(["text-indigo-500 whitespace-nowrap", { "text-size-18px leading-24px": __props.bigHeader }])
          }, [
            renderSlot(_ctx.$slots, "header")
          ], 2),
          createBaseVNode("p", _hoisted_4, [
            renderSlot(_ctx.$slots, "description", {}, () => [
              createBaseVNode("span", null, toDisplayString(__props.description), 1)
            ])
          ])
        ]),
        unref(slots).right ? (openBlock(), createElementBlock("div", _hoisted_5, [
          renderSlot(_ctx.$slots, "right")
        ])) : createCommentVNode("", true)
      ]);
    };
  }
});
const _sfc_main = /* @__PURE__ */ defineComponent({
  props: {
    gql: null,
    class: null,
    utmMedium: null
  },
  emits: ["success"],
  setup(__props, { emit }) {
    const props = __props;
    const { t } = useI18n();
    gql`
fragment CloudConnectButton on Query {
  currentProject {
    id
    config
  }
  ...CloudConnectModals
  ...LoginModal
}
`;
    const isLoginOpen = ref(false);
    const isProjectConnectOpen = ref(false);
    const isLoggedIn = computed(() => {
      var _a;
      return Boolean((_a = props.gql.cloudViewer) == null ? void 0 : _a.id);
    });
    function openConnection() {
      if (!isLoggedIn.value) {
        isLoginOpen.value = true;
      } else {
        isProjectConnectOpen.value = true;
      }
    }
    const cloudProjectId = computed(() => {
      var _a, _b, _c, _d;
      return (_d = (_c = (_b = (_a = props.gql) == null ? void 0 : _a.currentProject) == null ? void 0 : _b.config) == null ? void 0 : _c.find((item) => item.field === "projectId")) == null ? void 0 : _d.value;
    });
    return (_ctx, _cache) => {
      return openBlock(), createElementBlock(Fragment, null, [
        createVNode(_sfc_main$2, {
          class: normalizeClass(props.class),
          "prefix-icon": unref(isLoggedIn) ? unref(ConnectIcon) : unref(UserIcon),
          "prefix-icon-class": "icon-dark-white icon-light-transparent",
          onClick: openConnection
        }, {
          default: withCtx(() => [
            createTextVNode(toDisplayString(unref(isLoggedIn) ? unref(t)("runs.connect.buttonProject") : unref(t)("runs.connect.buttonUser")), 1)
          ]),
          _: 1
        }, 8, ["class", "prefix-icon"]),
        createVNode(_sfc_main$3, {
          modelValue: isLoginOpen.value,
          "onUpdate:modelValue": _cache[0] || (_cache[0] = ($event) => isLoginOpen.value = $event),
          gql: props.gql,
          "utm-medium": props.utmMedium,
          "show-connect-button-after-login": !unref(cloudProjectId),
          onConnectProject: _cache[1] || (_cache[1] = ($event) => isProjectConnectOpen.value = true)
        }, null, 8, ["modelValue", "gql", "utm-medium", "show-connect-button-after-login"]),
        isProjectConnectOpen.value ? (openBlock(), createBlock(_sfc_main$4, {
          key: 0,
          show: isProjectConnectOpen.value,
          gql: props.gql,
          "utm-medium": props.utmMedium,
          onCancel: _cache[2] || (_cache[2] = ($event) => isProjectConnectOpen.value = false),
          onSuccess: _cache[3] || (_cache[3] = ($event) => {
            isProjectConnectOpen.value = false;
            emit("success");
          })
        }, null, 8, ["show", "gql", "utm-medium"])) : createCommentVNode("", true)
      ], 64);
    };
  }
});
export {
  _sfc_main$1 as _,
  _sfc_main as a
};
